### PR TITLE
Generomak/core plasma

### PR DESCRIPTION
**Pull request by Mateasek**
This PR adds several functions providing Generomak Core plasma. The plasma profiles take 2 shapes. The first is a double quadratic, which is used for temperatures and some densities (H1+, C6+) and exponential decay for the rest. The 1D profiles are mapped into 3D using the generomak equilibrium, assuming constant values on magnetic surfaces. On default, the edge values of profiles are values of edge profiles on lcfs outer midplane. 

This PR is related to the issue #210

### CODE REVIEW

File: cherab/generomak/plasma/plasma.py [hunk 1]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+        equilibrium = load_equilibrium()
+        r = equilibrium.psin_to_r(1)
+        z = 0
+        lcfs_values = get_edge_profile_values(r, z)
+
+    # toroidal rotation profile
+    f1d_vtor = get_double_parabola(vtor_edge, vtor_core, vtor_convexity, vtor_concavity)
+
+    vpol_lcfs = 2e4
+    vpol_decay = 0.08
```
Comment (vsnever) [inline cherab/generomak/plasma/plasma.py:387]: `vpol_lcfs` and `vpol_decay` are passed as function arguments already.

File: cherab/generomak/plasma/plasma.py [hunk 2]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+            rest_mass = atomic_mass * spec_cherab.atomic_weight
+            species["composition"][name][chrg] = get_maxwellian_distribution(equilibrium, rest_mass=rest_mass,  **desc)
+        
+    return species.freeze()
+
+def get_core_plasma(distributions=None, atomic_data=None, parent=None, name="Generomak core plasma"):
+    """
+    Provides Generomak core plasma.
+
+    :param distributions: A dictionary of plasma distributions. Has to have the same for mas the
```
Comment (vsnever) [inline cherab/generomak/plasma/plasma.py:480]: `Has to have the same format as the ...`

File: cherab/generomak/plasma/plasma.py [hunk 3]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+                          dictionary returned by get_core_distributions. The default value
+                          is the value returned by the call get_core_distributions().
+    :param atomic_data: Instance of AtomicData, default isOpenADAS()
```
Comment (vsnever) [inline cherab/generomak/plasma/plasma.py:483]: Missing space in `default isOpenADAS()`.

File: cherab/generomak/plasma/plasma.py [hunk 4]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    :return: populated Plasma object
+    """
+    
+    # load Generomak equilibrium
+    equilibrium = load_equilibrium()
+
+    # create or check atomic_data
+    if atomic_data is not None:
+     if not isinstance(atomic_data, AtomicData):
+         raise ValueError("atomic_data has to be of type AtomicData")
```
Comment (vsnever) [inline cherab/generomak/plasma/plasma.py:495]: Indentation is not a multiple of 4.

File: cherab/generomak/plasma/plasma.py [hunk 5]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    # construct plasma primitive shape 
+    padding = 1e-3 #enlarge for safety
+    plasma_height = equilibrium.z_range[1] - equilibrium.z_range[0]
+    outer_column = Cylinder(radius=equilibrium.r_range[1], height=plasma_height)
+    inner_column = Cylinder(radius=equilibrium.r_range[0], height=plasma_height + 2 * padding)
+    inner_column.transform = translate(0, 0, -padding)
+    plasma_geometry = Subtract(outer_column, inner_column)
+
+    # coordinate transform of the plasma frame
+    geometry_transform = translate(0, 0, -outer_column.height / 2)
```
Comment (vsnever) [inline cherab/generomak/plasma/plasma.py:508]: I think this is incorrect. Should be:
```python
geometry_transform = translate(0, 0, equilibrium.z_range[0])
```
instead.

File: demos/generomak/plasma/plot_core_profiles.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+# setup temperature plot
+_, ax_t = plt.subplots()
+ax_t.set_title("Species Core Temperature Profiles")
+ax_t.set_xlabel("Psin")
+ax_t.set_ylabel("eV")
+
+# setup density plot
+_, ax_n = plt.subplots()
+ax_n.set_yscale("log")
+ax_n.tick_params(axis='both', which='both', labelsize=20)
```
Comment (vsnever) [inline demos/generomak/plasma/plot_core_profiles.py:38]: I don't understand why the font size for the tick labels should be increased here.

**Review by vsnever — CHANGES_REQUESTED** [review]
Hi @Mateasek,
Thank you very much for this. I suggest only minor changes.

I was thinking of a demo showing 2D profiles mapped to flux coordinates, but if you're going to add blended (core + edge) plasma later, a separate demo for 2D core profiles is not needed.

File: cherab/generomak/plasma/plasma.py [hunk 7]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    return initial_value * Exp1D((x - initial_position) * growth_rate)
+
+def get_maxwellian_distribution(equilibrium, f1d_density, f1d_temperature, f1d_vtor, f1d_vpol, f1d_vnorm, rest_mass):
+    """ Returns Maxwellian distribution for equilibrium mapped 1d profiles
+    
+    :param equilibrium: Instance of EFITEquilibrium
+    :param f1d_density: Function1D describing density profile.
+    :param f1d_temperature: Function1D describing temperature profile.
+    :param f1d_f1d_vtor: Function1D describing bulk toroidal rotation velocity profile.
+    :param f1d_f1d_vpol: Function1D describing bulk poloidal rotation velocity profile.
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py [hunk 8]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    :param f1d_vnom: Function1D describing bulk velocity normal to magnetic surfaces.
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py [hunk 9]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+                         nimp_convexity=2, nimp_concavity=4,
+                         nimp_decay=8,
+                         vtor_core=1e5, vtor_edge=1e4,
+                         vtor_convexity=2, vtor_concavity=4,
+                         vpol_lcfs=2e4, vpol_decay=0.08):
+    """
+    Returns dictionary of core profile functions and species descriptions
+
+    :param lcfs_values: Dictionary of profile values at the separatrix on outer midplane.
+                        The dictionary has to have the same for, mas the one returned by
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py [hunk 10]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+                        the function get_edge_profile_values. The default value is the
+                        dictionary returned by the call get_edge_profile_values for r, z
+                        on last closed flux surface on outer midplane.
+    :param ne_core: core electron density
+    :param ne_convexity: convexity of the electron density profile
+    :param ne_concavity: concavity of the electron density profile
+    :param te_core: core electron temperature 
+    :param te_convexity: convexity of the electron temperature profile
+    :param te_concavity: convexity of the electron temperature profile
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py [hunk 11]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    :param nh_core: core density of H1+
+    :param nh_convexity: convexity of H1+ density profile
+    :param nh_concavity: concavity of H1+ density profile
+    :param th_core: core H1+ temperature
+    :param th_convexity: convexity of H1+ density profile
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py (hunk 2 above)
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py (hunk 3 above)
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py (hunk 4 above)
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: How did I miss this? :-(.... Thanks

File: demos/generomak/plasma/plot_core_profiles.py (hunk 6 above)
Comment (Mateasek) [reply demos/generomak/plasma/plot_core_profiles.py]: This is an artefact of my pure eyesight and IDE setting. removed

**Comment by Mateasek** [comment]
Hi @vsnever ,

thanks for looking into this PR and for chewing through all the docstring typos. I accepted all the suggestions.

Regarding the mapping demo, it is a good idea and since we have the equilibrium in the generomak now, we can do it. Lets open an issue about it, to discus it there.

**Review by vsnever — APPROVED** [review]
Thanks, @Mateasek. I'm happy with this PR.

**Review by jacklovell — CHANGES_REQUESTED** [review]
Looks good, thanks @Mateasek. A few minor revisions from me.

File: cherab/generomak/plasma/plasma.py [hunk 12]
```diff
@@ -199,7 +201,7 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
 
     # create plasma composition list
     plasma_composition = []
-    for elem_name, elem_data in dists["composition"].items():
+    for _, elem_data in dists["composition"].items():
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: Thanks, changed

File: cherab/generomak/plasma/plasma.py [hunk 13]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    .. math:: x_{norm} = (x - xmin) / (xmax - xmin).
+
+    The returned function is decreasing and monotonous and its domain is [xmin, xmax].
+
+    :param v_min: The minimum value of the profile at xmax.
+    :param v_max: The maximum value of the profile at xmin.
+    :param xmin: The lower edge of the function domain. Defaults to 0.
+    :param xmax: The upper edge of the function domain Defaults to 1.
+    :param convexity: Controls the convexity of the profile in the lower values part of the profile.
+    :param concavity: Controls the concavity of the profile in the higher values part of the profile.
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: parameter docstrings reordered

File: cherab/generomak/plasma/plasma.py [hunk 14]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+                         nh_concavity=4, th_core=2.8e3,
+                         th_convexity=2, th_concavity=4,
+                         th0_fraction=0.8, nh0_decay=10,
+                         timp_core=2.7e3, timp_convexity=2,
+                         timp_concavity=4, nimp_core=5e17,
+                         nimp_convexity=2, nimp_concavity=4,
+                         nimp_decay=8,
+                         vtor_core=1e5, vtor_edge=1e4,
+                         vtor_convexity=2, vtor_concavity=4,
+                         vpol_lcfs=2e4, vpol_decay=0.08):
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: I added a function `get_core_profiles_arguments`, which returns a dictionary of core profile arguments. The function can be passed keyword arguments to change the default values. The list of arguments and their default values are in the function docstring. After a while I decided this is the best way to lower the number of function arguments while still having the description of core profile parameters somewhere.

**Comment by vsnever** [comment]
I think we should tweak the default concavity values for the core profiles to match the edge profiles better. I was trying to blend the core and the edge profiles together, and with the default `ne_concavity = 4` and `te_concavity = 3` I got this:
ne_concavity_4 0
te_concavity_3 0
But for the flatter core profiles, the consistency is much better. Here are the results for `ne_concavity = 0.2` and `te_concavity = 1`:
ne_concavity_0 2
te_concavity_1 0
I used the following function as a blending mask:
```python
blending_mask = equilibrium.map2d(Interpolator1DArray([0, 0.94, 1.0, 1.1], [1, 1, 0, 0], 'linear', 'none', 0))
```
I'll try tuning the concavities of other profiles as well.

**Comment by Mateasek** [comment]
Hi @vsnever,

could you please add plots of the 1D profiles? To me 0.2 and 1 seem a bit strange values.

**Comment by vsnever** [comment]
Hi @Mateasek,
Here are plots comparing normalised double-quadratic profiles with profiles calculated by ASTRA for a similar machine in H-mode. Profiles with lower concavities match the ASTRA profiles better, but can of course be further optimised.
normalised_density_profiles
normalised_temperature_profiles

**Comment by vsnever** [comment]
I found the optimal parameters of double-parabolic profiles by fitting the ASTRA profiles.
astra_profiles_fit
Here is the set of profile convexities, concavities and decays that is optimal in terms of matching ASTRA profiles for a similarly sized machine operating in H-mode and an existing SOLPS solution:
```
ne_convexity=1.09, ne_concavity=0.24,
te_convexity=2.35, te_concavity=1.26,
nh_convexity=1.09, nh_concavity=0.24,
th_convexity=1.00, th_concavity=0.82,
nimp_convexity=1.09, nimp_concavity=0.24,
timp_convexity=1.00, timp_concavity=0.82,
nh0_decay=20, nimp_decay=30
```
The most of the blended core+edge profiles look fine with these parameters:
ne_blend

te_blend
nh+_blend
th+_blend
nc4_blend
nc5_blend
nc6_blend
However, for H<sup>0</sup>, C<sup>0</sup>, C<sup>+</sup>, C<sup>2+</sup> and C<sup>3+</sup> density profiles, the value at LCFS has to be set manually to provide smooth transition between the core and the edge regions. I got nice results with the following parameters:
```
lcfs_values["composition"]["hydrogen"][0]["density"] = 1.e15
lcfs_values["composition"]["carbon"][2]["density"] = 1.e9
lcfs_values["composition"]["carbon"][3]["density"] = 1.e13
```
nh0_blend
nc2_blend
nc3_blend
As for the C<sup>0</sup> and C<sup>+</sup> densities, SOLPS simulation predicts they are zero at ψ<sub>n</sub><0.95.

I think that in most of the synthetic diagnostic demos for Generomak we will be using blended core and edge plasmas and so it is important that the core profile data is consistent with the existing edge solution.

**Comment by Mateasek** [comment]
I changed the constants to the ones suggested by @vsnever

File: cherab/generomak/plasma/plasma.py [hunk 15]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    :param th0_fraction: H0 temperature factor
+    :param nh0_decay: decay rate of H0 density profile
+    :param timp_core: core impurity density
+    :param timp_convexity: convexity of impurity temperature profile
+    :param timp_concavity: concavity of impurity temperature profile
+    :param nimp_core: core impurity density
+    :param nimp_convexity: convexity of impurity density profile
+    :param nimp_concavity: concavity of impurity density profile
+    :param nimp_decay: decay rate of impurity density profile (except bare nuclei)
+    :param vtor_core: toroidal rotation velocity at the edge m/s
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py [hunk 16]
```diff
@@ -215,3 +217,332 @@ def get_edge_plasma(atomic_data=None, parent=None, name="Generomak edge plasma")
...
+    :param vtor_edge: core toroidal rotation velocity m/s
+    :param vtor_convexity: convexity of the toroidal rotation profile
+    :param vtor_concavity: concavity of the toroidal rotation profile
+    :return: dictionary of Function1D profiles
+    """
```
[the remark this answers is not in the document]
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

File: cherab/generomak/plasma/plasma.py (hunk 1 above)
Comment (Mateasek) [reply cherab/generomak/plasma/plasma.py]: corrected

**Comment by Mateasek** [comment]
And now I fixed the doctring. So many typos :-(

**Review by jacklovell — APPROVED** [review]
Looks good to me now, happy to merge if @vsnever and @Mateasek agree that this is now ready.

It would be good to include a demo showing the integrated core and edge profiles that @vsnever has included in the discussion of this PR, illustrating how to blend them together to produce the 2D profiles. Happy for that to come in a second small PR though.

**Comment by jacklovell** [comment]
Ah, @Mateasek can you also add a changelog entry please?

**Comment by vsnever** [comment]
I think, it's ready. The default LCFS values for H<sup>0</sup>, C<sup>0</sup>, C<sup>+</sup>, C<sup>2+</sup> and C<sup>3+</sup> density profiles can be replaced later in a function that returns blended profiles. I'll create a separate PR with a `get_plasma()` function, which creates a plasma from blended core+edge profiles, and a demo showing the blending.

**Comment by Mateasek** [comment]
I think it would be better to accept the PR as is, and if needed, tweak the constants to match the blending.

@jacklovell I added changes to the changelog as requested.

**Pull request merged**